## Module 3 HW

In [12]:
! uv add pandas pyarrow google-cloud-bigquery db-dtypes google-cloud-bigquery-storage
import os, sys, time, requests
import pandas as pd
import db_dtypes
from google.cloud import bigquery, bigquery_storage
from google.api_core.exceptions import NotFound, Forbidden

Resolved 141 packages in 29ms
Audited 128 packages in 45ms


In [7]:
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "extras/gcp.json"

In [8]:
client = bigquery.Client.from_service_account_json(os.environ["GOOGLE_APPLICATION_CREDENTIALS"])

---

##### Tables:

In [25]:
# External Table.
ext = "abiding-bongo-485522-a4.nytaxi.external_yellow_tripdata"

In [26]:
# Materialized Table.
mat = "abiding-bongo-485522-a4.nytaxi.yellow_tripdata_non_partitioned"

---

### Question 1

What is the count of records for the 2024 Yellow Taxi Data? 

In [27]:
df = client.query(f"SELECT COUNT(*) FROM `{mat}`").to_dataframe()
print(df)

        f0_
0  20332093


Answer: `20,332,093`

---

### Question 2

What is the estimated amount of data that will be read when this query is executed on the External Table and the Table?

In [37]:
job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)

In [43]:
# For materialized table
query_job = client.query(
    f"SELECT COUNT(DISTINCT PULocationID) FROM `{mat}`",
    job_config=job_config
)
print(f"Materialized: {query_job.total_bytes_processed / (1024**3):.2f} GB")


Materialized: 0.15 GB


In [44]:
# For materialized table
query_job = client.query(
    f"SELECT COUNT(DISTINCT PULocationID) FROM `{ext}`",
    job_config=job_config
)
print(f"External: {query_job.total_bytes_processed / (1024**3):.2f} GB")


External: 0.00 GB


Answer: `0 MB for the External Table and 155.12 MB for the Materialized Table`

---

### Question 3

Why are the estimated number of Bytes different?

Answer:

---

### Question 4

How many records have a fare_amount of 0?

Answer:

---

### Question 5

What is the best strategy to make an optimized table in Big Query if your query will always filter based on tpep_dropoff_datetime and order the results by VendorID (Create a new table with this strategy)

Answer:

___

### Question 6

Write a query to retrieve the distinct VendorIDs between tpep_dropoff_datetime 2024-03-01 and 2024-03-15 (inclusive). Use the materialized table you created earlier in your from clause and note the estimated bytes. Now change the table in the from clause to the partitioned table you created for question 5 and note the estimated bytes processed. What are these values?

Answer:

---

### Question 7

Where is the data stored in the External Table you created?

Answer:

---

### Question 8

It is best practice in Big Query to always cluster your data:

Answer:

---

### Question 9

Write a `SELECT count(*)` query FROM the materialized table you created. How many bytes does it estimate will be read? Why?

Answer:

---